### PR TITLE
Perform the initial tagging for a file as fast as possible so that that the file appears 'complete' immediately after opening.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Mitigates: https://github.com/dotnet/roslyn/issues/18376

Note: no matter what, there may be some delay on some tags.  This is because computing tags actually does take time.  And either we block the editor on computing the tags (i.e. synchronous tag computation, which could def slow down file open), or we're async, but that means you can observe a period of time without the tags.

This also only affects initial tag reporting when a tagger is created.  It does not *by design* change any of our delays and whatnot reporting tags after edits or caret moving.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Experience opening files and using the provisional tab now is like:

fastinitialdocumenttagging

fastinitialdocumenttagging1

**Comment by CyrusNajmabadi** [comment]
Using provisional tabs is now:

fastinitialdocumenttagging1

**Comment by CyrusNajmabadi** [comment]
@rchande How are we getting along with perf tests?  this is something that could definitely trigger perf regression markers.  I'd like to know about that ahead of time if possible.  Thanks!

**Comment by lorcanmooney** [comment]
Running from this branch causes a `NullReferencException` in src\EditorFeatures\Core\Tagging\AbstractAsynchronousTaggerProvider.Tagger.cs at line 89

```csharp
                        var collection = new NormalizedSnapshotSpanCollection(
                            tags.GetSpans(_subjectBuffer.CurrentSnapshot).Select(ts => ts.Span));
```

In the above code, `tags` is null.

**Comment by CyrusNajmabadi** [comment]
Interesting.  I don't repro that myself.  but i'll maket he code resilient to it as it can certainly happen if the tagger is created before the tagsource finishes computing.

**Comment by lorcanmooney** [comment]
This doesn't seem to reliably fix the problem for the error squiggles. :-(
Maybe 1-in-5 they show up immediately, but most of the time they're still slow to appear.

**Comment by CyrusNajmabadi** [comment]
> Maybe 1-in-5 they show up immediately, but most of the time they're still slow to appear.

diagnostics are another system (and quite complex).  I'm not trying to do anything in that space yet.  Note: it may truly be hte case (even with FSA) that these diagnostics are just gone and need to be recomputed for the document in question.

**Comment by lorcanmooney** [comment]
> I'm not trying to do anything in that space yet. 

Sorry, my mistake.

> ...these diagnostics are just gone and need to be recomputed for the document in question.

I did notice that every time I open a document the little "..." appears in the error list. Maybe it is recomputing something.

**Comment by CyrusNajmabadi** [comment]
Technically when a file is opened, it changes.  it has been interpreted by the editor and may not represent teh same bytes that we got when we read the contents in from disk.  For example, it may have normalized newlines, or picked an entirely different encoding than we picked.  It may even have been checked out from source control on open, and may be vastly different.  It's an open item that we use things like file hashes to know if we're really processing the same file.

however, we have to conservatively assume that we're not as the file really may be different.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs [hunk 1]
```diff
@@ -196,20 +199,32 @@ private bool UpToDate
...
 
             public void RegisterNotification(Action action, int delay, CancellationToken cancellationToken)
-            {
-                _notificationService.RegisterNotification(action, delay, _asyncListener.BeginAsyncOperation("TagSource"), cancellationToken);
-            }
+                => _notificationService.RegisterNotification(action, delay, _asyncListener.BeginAsyncOperation("TagSource"), cancellationToken);
+
+            private void RecalculateTagsOnChanged(TaggerEventArgs e)
+                => RecalculateTagsOnChanged_DoNotCallDirectly(
+                    (int)e.Delay.ComputeTimeDelay().TotalMilliseconds, cancellable: false);
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs]: Shouldn't this one be cancellable?

Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs]: OH DEAR LORD

**Comment by AdamSpeight2008** [comment]
Can we cache the tagging information? so it doesn't have to be recomputed on file opening. Provided that the source hasn't changed.

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs [hunk 2]
```diff
@@ -108,8 +115,22 @@ private sealed partial class TagSource : ForegroundThreadAffinitizedObject
...
+                // possible, without kicking the work  down the road.
+                var initialTagsCancellationToken = _initialComputationCancellationTokenSource.Token;
+                if (_workQueue.IsForeground())
+                {
+                    RecomputeTagsForeground(cancellationTokenOpt: initialTagsCancellationToken);
+                }
+                else
+                {
+                    RegisterNotification(
+                        () => RecomputeTagsForeground(cancellationTokenOpt: initialTagsCancellationToken),
```
Comment (AdamSpeight2008) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs]: Can this `RecomputeTagsForeground` be done in the background?

**Comment by CyrusNajmabadi** [comment]
> Can we cache the tagging information? so it doesn't have to be recomputed on file opening.

At the cost of memory, potentially.  We could store to disk.  But then you're bringing more IO into the mix, and that's also potentially quite slow.

> Provided that the source hasn't changed.

There's the rub.  How do you determine the source hasn't changed?

File: src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource.cs]: Not without a large rewrite of tagging.  The UI thread acts as a sync point for us to ensure proper state introspection and mutation.  We try to keep that work to a minimum, doing the actual computation on the BG thread.  The UI thread is used to just store state in a consistent fashion so we can not lock when things like GetTags are called.

**Comment by CyrusNajmabadi** [comment]
@Pilchie I'm not going to check this in without validation from RPS/ravi that there is no measured problems for this change.

**Comment by AdamSpeight2008** [comment]
Visual Studio can already detect external file changes, when the project file is open. 
We just need a system that monitors the files when they are closed and VS is running. So it can invalidate the tagger cache for that file, those prompting a full recompute.

**Comment by CyrusNajmabadi** [comment]
@AdamSpeight2008 that only handles when the file is changed on disk.  There are cases of file content changes without the file changing on disk.  For example, roslyn may read in the file with one encoding, and VS may choose another encoding.  Things like newlines may be normalized.  Files may be checked out of source control.

Without actually validating that our bytes are exactly the same as the interpretation of VS's you can get different results.  We have to be conservative and assume the file may have changed on open until some sort of API can be provided that proves the contents are the same.

Right now the only API for that is actually checking the file contents (or some sort of good hash).  But that's expensive, meaning you still end up with a delay on file open.

**Comment by CyrusNajmabadi** [comment]
Also, as i mentioned, what you're discussing is potentially quite expensive.  Caches are not free (either in memory or disk space), and we already use a lot to at least speed up operations that are quite slow.  Taking that sort of hit for operations that are already generally quite fast is not necessarily a good idea.

**Comment by rchande** [comment]
@CyrusNajmabadi I will put "Open File" on my backlog of tests to get running. I believe every RPS test has an "open file" scenario so we'll get some coverage from there.

**Comment by CyrusNajmabadi** [comment]
Thanks!  I will wait for that to happen before sending this in.  Please let me know when it does @rchande .  Thanks!

**Comment by CyrusNajmabadi** [comment]
@rchande Do we have a perf test for this available yet?

**Comment by rchande** [comment]
test internal-perf please

**Comment by rchande** [comment]
@CyrusNajmabadi The Jenkins trigger seems not to be working. Let me look into that.

**Comment by CyrusNajmabadi** [comment]
Thanks @rchande .  Once we can test this i can get this merged in.

**Comment by CyrusNajmabadi** [comment]
test internal-perf please

**Comment by CyrusNajmabadi** [comment]
retest this please

**Comment by CyrusNajmabadi** [comment]
test internal-perf please

**Comment by rchande** [comment]
test internal-perf please

**Comment by CyrusNajmabadi** [comment]
RPS results show no degradation.  Merging into 15.6 so people can try this out.

**Pull request merged**